src/Storages/MergeTree/MergeTask.cpp

```diff
@@ -518,7 +517,23 @@ bool MergeTask::VerticalMergeStage::prepareVerticalMergeForAllColumns() const
             "of bytes written to rows_sources file ({}). It is a bug.",
             sum_input_rows_exact, input_rows_filtered, cleanedup_rows_count, rows_sources_count);
 
-    ctx->rows_sources_read_buf = std::make_unique<CompressedReadBufferFromFile>(ctx->tmp_disk->readFile(fileName(ctx->rows_sources_file->path())));
+    /// TemporaryDataOnDisk::createRawStream returns WriteBufferFromFile implementing IReadableWriteBuffer
+    /// and we expect to get ReadBufferFromFile here.
+    /// So, it's relatively safe to use dynamic_cast here and downcast to ReadBufferFromFile.
+    auto * wbuf_readable = dynamic_cast<IReadableWriteBuffer *>(ctx->rows_sources_uncompressed_write_buf.get());
+    auto reread_buf = wbuf_readable ? wbuf_readable->tryGetReadBuffer() : nullptr;
+    if (!reread_buf)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot reread temporary file {}", ctx->rows_sources_uncompressed_write_buf->getFileName());
+    auto * reread_buffer_raw = dynamic_cast<ReadBufferFromFile *>(reread_buf.get());
+    if (!reread_buffer_raw)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot reread temporary file {}", ctx->rows_sources_uncompressed_write_buf->getFileName());
+    /// Move ownership:
+    /// First, release ownership from unique_ptr to base type.
+    reread_buf.release(); /// NOLINT(bugprone-unused-return-value): we already have the pointer value in `reread_buffer_raw`
+    /// Then, move ownership to unique_ptr to concrete type.
+    std::unique_ptr<ReadBufferFromFile> reread_buffer_from_file(reread_buffer_raw);
```

Comment:
I do not think this is allowed, now you have shared pointer in `ctx->rows_sources_uncompressed_write_buf` and unique pointer `reread_buffer_from_file` pointing to the same object.

Reply:
The `rows_sources_uncompressed_write_buf` is an instance of `WriteBufferFromFileBase` type, but here I transfer ownership of `reread_buf` which is `ReadBuffer` obtained from `tryGetReadBuffer` and is a new independent object.
Additionally, in this PR I changed return type of `tryGetReadBuffer` to `std::unique_ptr<ReadBuffer>`, so we own pointer in `reread_buf`. Thus, performed operation is like `dynamic_pointer_cast` but for `unique_ptr`s.

Alternatively, we can try to change `IReadableWriteBuffer` and add couple of methods there that returns concrete `ReadBufferWhatever` instead of `std::unique_ptr<ReadBuffer>`. This would require a bit more changes across the codebase but should be doable.

Reply:
> and is a new independent object.

ah, indeed, it was not obvious unless opening the code of `tryGetReadBuffer` and `getReadBufferImpl`. 
May be it is better to make `tryGetReadBuffer` method return `unique_ptr`? As it seems it is always a bew object. This will solve this issue with converting shared ptr to unique ptr here in MergeTask.

Reply:
Ah, yeah sorry I confused this part. Let's leave it as it is then.